ci/cron/src/Main.hs

```diff
@@ -0,0 +1,254 @@
+    putStrLn $ "Building latest docs: " <> latest
+    shell_ $ "git checkout v" <> latest
+    robustly_download_nix_packages
+    shell_ "bazel build //docs:docs"
+    shell_ $ "tar xzf bazel-genfiles/docs/html.tar.gz --strip-components=1 -C" <> path
+    -- Not going through Aeson because it represents JSON objects as unordered
+    -- maps, and here order matters.
+    let versions_json = versions
+                        |> map (\s -> "\"" <> s <> "\": \"" <> s <> "\"")
+                        |> List.join ", "
+                        |> \s -> "{" <> s <> "}"
+    writeFile (path <> "/versions.json") versions_json
+    shell_ $ "mkdir -p  " <> path <> "/" <> latest
+    shell_ $ "tar xzf bazel-genfiles/docs/html.tar.gz --strip-components=1 -C " <> path <> "/" <> latest
+    Foldable.for_ (tail versions) $ \version -> do
+        putStrLn $ "Building older docs: " <> version
+        shell_ $ "git checkout v" <> version
+        robustly_download_nix_packages
+        shell_ "bazel build //docs:docs"
+        shell_ $ "mkdir -p  " <> path <> "/" <> version
+        shell_ $ "tar xzf bazel-genfiles/docs/html.tar.gz --strip-components=1 -C" <> path <> "/" <> version
+    shell_ $ "git checkout " <> cur_sha
+
+check_s3_versions :: Set.Set String -> IO Bool
+check_s3_versions gh_versions = do
+    temp <- shell "mktemp"
+    shell_ $ "aws s3 cp s3://docs-daml-com/versions.json " <> temp
+    s3_raw <- shell $ "cat " <> temp
+    case JSON.decode $ LBS.fromString s3_raw of
+      Just s3_json -> return $ (docs_versions s3_json) == gh_versions
+      Nothing -> Exit.die "Failed to get versions from s3"
+
+push_to_s3 :: String -> IO ()
+push_to_s3 doc_folder = do
+    putStrLn "Pushing new versions file first..."
+    shell_ $ "aws s3 cp " <> doc_folder <> "/versions.json s3://docs-daml-com/versions.json --acl public-read"
+    putStrLn "Pushing to S3 bucket..."
+    shell_ $ "aws s3 sync " <> doc_folder
+             <> " s3://docs-daml-com/"
+             <> " --delete"
+             <> " --acl public-read"
+             <> " --exclude '*.doctrees/*'"
+             <> " --exclude '*.buildinfo'"
+    putStrLn "Refreshing CloudFront cache..."
+    shell_ $ "aws cloudfront create-invalidation"
+             <> " --distribution-id E1U753I56ERH55"
+             <> " --paths '/*'"
+
+data BlogId = BlogId { blog_id :: Integer } deriving Show
+instance JSON.FromJSON BlogId where
+    parseJSON = JSON.withObject "BlogId" $ \v -> BlogId
+        <$> v JSON..: (Text.pack "id")
+
+data SubmitBlog = SubmitBlog {
+                      body :: String,
+                      date :: Integer,
+                      summary :: String,
+                      version :: String
+                  } deriving Show
+instance JSON.ToJSON SubmitBlog where
+    toJSON SubmitBlog{body, date, summary, version} =
+    -- content_group_id and blog_author_id reference existing items in HubSpot
+        JSON.object ["name" JSON..= ("Release of DAML SDK " <> version),
```

Comment:
I would import `.=` unqualified. It’s sufficiently common that the costs of not littering the code with qualified imports are lower than the costs of learning the operator imho.

Reply:
I personally think the cost of having import qualifiers is way lower than the cost of not having them. (You have to learn the operator either way.)